// presage/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.series = exports.reduce = exports.promiseWithCallback = exports.parallel = exports.mapSeries = exports.map = exports.filterSeries = exports.filter = exports.eachSeries = undefined;

var _eachSeries = require('./each-series');

var _eachSeries2 = _interopRequireDefault(_eachSeries);

var _filter = require('./filter');

var _filter2 = _interopRequireDefault(_filter);

var _filterSeries = require('./filter-series');

var _filterSeries2 = _interopRequireDefault(_filterSeries);

var _map = require('./map');

var _map2 = _interopRequireDefault(_map);

var _mapSeries = require('./map-series');

var _mapSeries2 = _interopRequireDefault(_mapSeries);

var _parallel = require('./parallel');

var _parallel2 = _interopRequireDefault(_parallel);

var _promiseWithCallback = require('./promise-with-callback');

var _promiseWithCallback2 = _interopRequireDefault(_promiseWithCallback);

var _reduce = require('./reduce');

var _reduce2 = _interopRequireDefault(_reduce);

var _series = require('./series');

var _series2 = _interopRequireDefault(_series);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.eachSeries = _eachSeries2.default;
exports.filter = _filter2.default;
exports.filterSeries = _filterSeries2.default;
exports.map = _map2.default;
exports.mapSeries = _mapSeries2.default;
exports.parallel = _parallel2.default;
exports.promiseWithCallback = _promiseWithCallback2.default;
exports.reduce = _reduce2.default;
exports.series = _series2.default;
exports.default = {
    eachSeries: _eachSeries2.default,
    filter: _filter2.default,
    filterSeries: _filterSeries2.default,
    map: _map2.default,
    mapSeries: _mapSeries2.default,
    parallel: _parallel2.default,
    promiseWithCallback: _promiseWithCallback2.default,
    reduce: _reduce2.default,
    series: _series2.default
};
