// igdb-api-node/lib/perform-request.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _requestPromise = require('./request-promise');

var _requestPromise2 = _interopRequireDefault(_requestPromise);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (url, apiService) {
    var options = {
        headers: {
            Accept: 'application/json'
        },
        url: url
    };

    options.headers[apiService.keyHeader] = apiService.key;

    return (0, _requestPromise2.default)(options).then(function (result) {
        return new _promise2.default(function (resolve, reject) {
            if (result.response.statusCode !== 200) {
                reject(new Error('HTTP Status ' + result.response.statusCode + ' - ' + options.url));
                return;
            }

            try {
                resolve({
                    body: JSON.parse(result.body),
                    headers: result.response.headers,
                    scrollCount: result.response.headers['x-count'] || result.response.headers['X-Count'],
                    scrollUrl: result.response.headers['x-next-page'] || result.response.headers['X-Next-Page'],
                    url: options.url
                });
            } catch (error) {
                reject(error);
            }
        });
    });
};
